MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,961 @@
+authors: Deepthi Kumar (deepthiskumar), Paul Steckler (@psteckler), Brett Carter (@carterbrett), Brandon Kase (@bkase)
+discussions-to:
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-02-14
+---
+
+## Abstract
+
+This MIP proposes adding programmable smart contracts, called zkApps, to the Mina protocol.
+
+The reference implementation is located on the `develop` branch of the
+MinaProtocol/mina repo.
+
+## Motivation
+
+The zkApps protocol aims to add a more fleshed out smart contract layer to Mina
+to make it much more convenient and flexible to add custom applications.
+These changes extend the potential of zero-knowledge cryptography by enabling
+the following characteristics while preserving the succinctness of Mina:
+
+  **General programmability** - able to execute and settle arbitrary programs,
+  not constrained to a particular VM model, but instead is flexibly designed to
+  support any number of execution models, VM-like or otherwise
+
+  **Programmable privacy** - the privacy of inputs to smart contracts and
+  their state can be programmed by their developers
+
+  **Constant in verification time** - individual transactions are executed, or
+  more accurately “proven” asynchronously off-chain and verified on-chain in
+  time proportional to the account updates, independent of computational
+  complexity of any individual proof inside of each account update
+
+This MIP represents the first iteration of such a protocol.
+
+## Specification
+
+The following description summarizes the planned changes to the protocol. Not
+all technical details are described here.
+
+### Ledger account changes
+
+Recall that accounts are identified by a public key and token id pair.
+
+To accommodate zkApps, Mina ledger accounts have the following new field:
+
+- `token_symbol`: a string, up to six characters long, that can provide a name
+     for a `token_id` whose owner is the public key and token id in this
+     account. In the existing mainnet, only the default MINA `token_id` is
+     used. New `token_id`s are derived from the the addresses of the
+     smart contract that defines them.
+
+The following account field is removed:
+
+- `token_permissions`: this field was meant for non-default tokens that were
+never introduced
+
+The `snapp` field in accounts was never instantiated and is
+replaced by the `zkapp` field.
+
+The `zkapp` field is a record with the following fields:
+
+- `app_state` : an array of eight field elements, the current state of a zkApp
+associated with the account
+- `verification_key` : an optional key to use when verifying transaction
+SNARKs for zkApps using this account; when not provided, the account is a
+typical Mina account that supports payments.
+- `zkapp_version` : a nonnegative integer, the zkApps version that determined
+the `app_state`
+- `action_state` : an array of five field elements, each representing an
+"action" (see below) contained in a zkApp
```

Comment:
```suggestion
- `action_state` : an array of five field elements, each representing an
"action" contained in a zkApp. For details, see [actions](#actions).

```

Reply:
afaik, in-document links work to headings, so I've added the ####Actions heading